_posts/2018-01-30-quicli.md

```diff
@@ -0,0 +1,294 @@
+But how exactly do I want to to this?
+Thanks to Garrett's questions and comments
+I now have an answer to this.
+
+## What do I want to quicli to be
+
+(The following are basically quotes from [#19].)
+
+My goal is to provide
+an opinionated set
+of convenience functions
+to quickly build CLI apps.
+
+In its implementation,
+you get what I'd call
+"a small framework around your main function":
+quicli's `main!` macro set up some basic things,
+and its prelude gives a few tools
+that are commonly required in CLI apps.
+These tools are either re-exports from existing Rust crates,
+or abstractions around library functions that are more ergonomic
+that better fit the CLI use-case
+(as well as my ergonomic requirements).
+
+What I want you to end up with is code that is
+concise,
+well-structured and boilerplate-free,
+and production-ready.
+Okay, that's nothing new; we all want that.
+But what does this even mean for a CLI app?
+
+- concise
+
+	Ideally: You type one line to get one thing done.
+	If you want to read a file
+	and see if its text contains the word "foobar",
+	you can do:
+
+	```rust
+	let content = read_file("./input.txt")?;
+	let has_foobar = content.contains("foobar");
+	```
+	
+	You want to read a file, I give you a `read_file` function.
+	This skips allocating a String as buffer, opening File, and calling `read_to_string` on that file.
+	And while reading a file to a string this way may not be the most performant option,
+	it is very concise.
+	
+	Another aspect of conciseness:
+	Instead of using `.unwrap` to deal with errors,
+	quicli's `main!` allows you to use `?` to propagate errors,
+	and shows human-readable messages when your program exited with an error state.
+	Similarly, instead of
+	writing `return Err(SomeType::new("message"))`,
+	by using quicli you also get [failure]'s `bail!` and `ensure!`.
+	
+- well-structured and boilerplate-free
+
+	I just repeatedly wrote that this is a framework for _small_ code bases.
+	Nevertheless, it makes sense to structure your code well.
+	Actually, structuring your code in a clever way helps you keep the code size small.
+	
+	For example, this is how you parse CLI arguments (incl. type checks):
+	
+	```rust
+	/// Get first n lines of a file
+	#[derive(Debug, StructOpt)]
+	struct Cli {
+	    /// How many lines to get
+	    #[structopt(long = "count", short = "n", default_value = "3")]
+        count: usize,
+        /// The file to read
+	    file: String,
+    }
+    ```
+
+	See how you didn't write any code to convert whatever followed `--count` to an integer?
+	By the way:
+	Adding doc comments to your CLI arguments structure is not just for when you write the code,
+	structopt will also show them in a `--help`.
+
+- production-ready
+
+	In my experience,
+	the best way to ensure something ends up being used in production for years to come
+	is to call it a prototype.
+	So, I always want to get some basic stuff that will make using even the smallest tool less of a pain.
```

Comment:
You can just remove `So,` here. It's a bit of filler in English that lessens the impact of your words. You ALWAYS want to do this, not as a "well sometimes" kind of feeling So implies